pkg/chart/chart.go

```diff
@@ -222,13 +253,42 @@ func (t *Testing) processCharts(action func(chart string, valuesFiles []string)
 		TestResults:    results,
 	}
 
+	if t.config.Upgrade {
+		changedParentDirs := map[string]bool{}
+		for _, dir := range charts {
+			changedParentDirs[path.Dir(dir)] = true
+		}
+		for dir := range changedParentDirs {
+			// Check for uncommitted changes that would be lost when checking out older revision
+			clean, _ := t.git.IsDirClean(dir)
+			if !clean {
+				return results, fmt.Errorf("Directory %s has uncommitted changes", dir)
+			}
+			// Set contents of charts directory to the target branch
+			t.git.CheckoutDir(dir, t.mergeBase)
+			// Copy charts directory contents to a temp directory
+			prevRevisionDir := tempChartParentDir(dir)
+			copy.Copy(dir, prevRevisionDir)
+			// Reset charts directory to last commit
+			t.git.CheckoutDir(dir, "HEAD")
+			// Schedule cleanup of untracked temp directory
+			defer t.git.CleanDir(prevRevisionDir)
```

Comment:
`defer` should not be used inside a loop.

Reply:
To me the defer feels more maintainable since we schedule the cleanup action right after creating the temp directory, and we would need to do some additional bookkeeping for handling early returns otherwise.

Memory overhead here should be very low since we're only keeping track of the `prevRevisionDir` string, and disk usage scales linearly (if disk is a concern we'll need to do a little additional refactoring).

Reply:
If I understand this correctly, when using defer inside a loop we should wrap it in either an anonymous or a named function. See my note about previous version(s) above… this may cause the loop to pile up a lot more.